File: pkg/bt/deserialize.go
package bt

import "encoding/json"

// TreeFile represents a JSON file with a named behavior tree.
// The Blackboard field holds initial data populated when the tree is loaded.
type TreeFile struct {
	Name       string         `json:"name"`
	Blackboard map[string]any `json:"blackboard,omitempty"`
	Tree       json.RawMessage `json:"tree"`
}

// DeserializeTree parses a TreeFile JSON document and builds the node tree.
// Returns the root Node and the initial blackboard data (may be nil).
// The caller should create a Blackboard from the returned data and pass it
// via ContextWithBlackboard before ticking.
func DeserializeTree(data []byte, reg *FactoryRegistry) (Node, map[string]any, error) {
	var tf TreeFile
	if err := json.Unmarshal(data, &tf); err != nil {
		return nil, nil, err
	}

	var rootCfg NodeConfig
	if err := json.Unmarshal(tf.Tree, &rootCfg); err != nil {
		return nil, nil, err
	}

	// Merge top-level blackboard initial data into the root config
	if len(tf.Blackboard) > 0 {
		if rootCfg.Blackboard == nil {
			rootCfg.Blackboard = make(map[string]any)
		}
		for k, v := range tf.Blackboard {
			if _, exists := rootCfg.Blackboard[k]; !exists {
				rootCfg.Blackboard[k] = v
			}
		}
	}

	node, err := reg.Build(rootCfg)
	if err != nil {
		return nil, nil, err
	}
	return node, tf.Blackboard, nil
}

// DeserializeNode parses a single NodeConfig JSON and builds it.
func DeserializeNode(data []byte, reg *FactoryRegistry) (Node, error) {
	var cfg NodeConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		return nil, err
	}
	return reg.Build(cfg)
}

File: bt_service/factory/deserialize.py
"""JSON deserialization — parses tree JSON into a Node tree.

Mirrors Go pkg/bt/deserialize.go.
"""

import json
from typing import Any

from bt_service.core.node import Node, Status
from bt_service.factory.registry import FactoryRegistry
from bt_service.factory.config import NodeConfig, TreeFile


def deserialize_tree(data: str | bytes, reg: FactoryRegistry) -> tuple[Node, dict[str, Any] | None]:
    """Parse a TreeFile JSON string and build the node tree.

    Returns (root Node, initial_blackboard_data).
    The caller should create a Blackboard from initial data before ticking.
    """
    if isinstance(data, bytes):
        data = data.decode("utf-8")

    d = json.loads(data)
    tf = TreeFile.from_dict(d)

    root_cfg = tf.tree or {}
    if not root_cfg:
        raise ValueError("tree file has no 'tree' key")

    # Merge top-level blackboard into root config
    if tf.blackboard:
        root_bb = root_cfg.get("blackboard") or {}
        merged = dict(tf.blackboard)
        merged.update(root_bb)  # root keys override top-level
        root_cfg["blackboard"] = merged

    node = reg.build(root_cfg)
    return node, tf.blackboard


def deserialize_node(data: str | bytes, reg: FactoryRegistry) -> Node:
    """Parse a single NodeConfig JSON string and build the node."""
    if isinstance(data, bytes):
        data = data.decode("utf-8")
    cfg = json.loads(data)
    return reg.build(cfg)
